integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/BlazeCriteriaBuilderRenderer.java

```diff
@@ -0,0 +1,1308 @@
+                                    baseCriteriaBuilder.union() : baseCriteriaBuilder.unionAll();
+                            subQuery = (SubQueryExpression<?>) unionOperation.getArg(1);
+                            ((SelectCTECriteriaBuilder<?>) serializeSubQuery(recursiveCriteriaBuilder, subQuery)).end();
+                        } else {
+                            FullSelectCTECriteriaBuilder<?> cteBuilder = criteriaBuilder.with(cteType);
+
+                            Object result = serializeSubQuery(cteBuilder, subQueryExpression);
+
+                            if (result instanceof FinalSetOperationCTECriteriaBuilder) {
+                                ((FinalSetOperationCTECriteriaBuilder<?>) result).end();
+                            } else if (result instanceof FullSelectCTECriteriaBuilder) {
+                                ((FullSelectCTECriteriaBuilder<?>) result).end();
+                            }
+                        }
+                        return;
+                    case WITH_RECURSIVE_COLUMNS:
+                    case WITH_COLUMNS:
+                        cteAliases = args.get(1).accept(new CteAttributesVisitor(), new ArrayList<Path<?>>());
+                        return;
+                    default:
+                        break;
+                }
+            }
+
+            // JPQLSerializer calls serialize transitively,
+            if (operator == Ops.EXISTS && args.get(0) instanceof SubQueryExpression) {
+                append("EXISTS (");
+                renderSubQueryExpression(args.get(0));
+                append(")");
+                return;
+            }
+
+            super.visitOperation(type, operator, args);
+        }
+
+        @Override
+        public Void visit(Operation<?> expr, Void context) {
+            if (expr.getOperator() instanceof JPQLNextOps) {
+                switch ((JPQLNextOps) expr.getOperator()) {
+                    case SET_UNION:
+                    case SET_UNION_ALL:
+                    case SET_INTERSECT:
+                    case SET_INTERSECT_ALL:
+                    case SET_EXCEPT:
+                    case SET_EXCEPT_ALL:
+                    case LEFT_NESTED_SET_UNION:
+                    case LEFT_NESTED_SET_UNION_ALL:
+                    case LEFT_NESTED_SET_INTERSECT:
+                    case LEFT_NESTED_SET_INTERSECT_ALL:
+                    case LEFT_NESTED_SET_EXCEPT:
+                    case LEFT_NESTED_SET_EXCEPT_ALL:
+                        renderSubQueryExpression(expr);
+                        return null;
+                    default:
+                        break;
+                }
+            }
+
+            return super.visit(expr, context);
+        }
+
+    }
+
+    /**
+     * Visitor for rendering window expressions.
+     *
+     * @param <X> Concrete window builder
+     */
+    private class WindowContainerBuilderDefaultVisitorImpl<X extends WindowContainerBuilder<X>> extends DefaultVisitorImpl<X, Object> {
```

Comment:
No, it needs the context of the query rendering for registering constants and parameters.

Reply:
AFAICS you only need the `renderExpression` method which I suggested could be _moved_ to the serializer class. Then you could pass just the serializer unless I didn't miss something.

Reply:
We probably could use a plain `JPQLNextSerializer` to serialize the expressions used in the window clause, because I _think_ subqueries are not allowed there. I am not even sure of that. If they are allowed, it gets a lot harder.

I don't think that we can easily move the `BlazeCriteriaBuilderRenderer#renderExpression` to `JPQLNextSerializer`. We need special handling for some expression types. Nested CTE's, SubqueryInitiators, QueryFlags, window functions, all get in the way here. For example, the `JPQLNextSerializer` just writes the entire window to an expression string, but the renderer has to use a visitor to it can call the right builder methods.

I can add a whole lot of context to the JPQLNextSerializer that can be pushed and popped, cte's, subqueries, windows, CTE binds, so that the `BlazeCriteriaBuilderRenderer` can retrieve it from there, but I think that would just eventually mean that we integrate `BlazeCriteriaBuilderRenderer` with `JPQLNextSerializer` into a single class. This is actually something I had actually done before but I really wanted to pull out again because the `BlazeCriteriaBuilderRenderer`  can't satisfy the regular `JPQLSerializer` interface and would publically expose a lot of unwanted methods.

Reply:
Ok fine, it was just an idea.